qiling/loader/pe.py

```diff
@@ -629,16 +629,152 @@ def init_ki_user_shared_data(self):
-        kusd_obj.NtMinorVersion = sysconf.getint('minorVersion')
-        kusd_obj.KdDebuggerEnabled = 0
-        kusd_obj.NXSupportPolicy = 0        # NX_SUPPORT_POLICY_ALWAYSOFF
+        kusd_obj.ImageNumberLow     = 0x014c                # IMAGE_FILE_MACHINE_I386
+        kusd_obj.ImageNumberHigh    = 0x8664                # IMAGE_FILE_MACHINE_AMD64
+        kusd_obj.NtSystemRoot       = self.ql.os.windir
+        kusd_obj.NtProductType      = sysconf.getint('productType')
+        kusd_obj.NtMajorVersion     = sysconf.getint('majorVersion')
+        kusd_obj.NtMinorVersion     = sysconf.getint('minorVersion')
+        kusd_obj.KdDebuggerEnabled  = 0
+        kusd_obj.NXSupportPolicy    = 0                     # NX_SUPPORT_POLICY_ALWAYSOFF
 
         self.ql.os.KUSER_SHARED_DATA = kusd_obj
+    
+    def init_kpcr(self):
+        '''
+        Initialisation function for KPCR structure. 
+        
+        This structure's pointer should always be at gs:[0x18]
+        '''
+
+        sysconf = self.ql.os.profile['SYSTEM']
+        osconf  = self.ql.os.profile[f'OS{self.ql.arch.bits}']
+
+        kpcr_addr = osconf.getint('KPCR')
+        kpcr_struct = KPCR
+        self.ql.mem.map(kpcr_addr, self.ql.mem.align_up(kpcr_struct.sizeof()), info='[kpcr]')
+
+        # Get address for KPRCB structure
+        kprcb_addr = osconf.getint('KPRCB')
+
+        # Initialize an instance with a few key fields
+        # @TODO: initialize StackBase & StackLimit
+        kpcr_obj = kpcr_struct.volatile_ref(self.ql.mem, kpcr_addr)
+        kpcr_obj.MajorVersion = sysconf.getint('majorVersion')
+        kpcr_obj.MinorVersion = sysconf.getint('minorVersion')
+        kpcr_obj.CurrentPrcb = kprcb_addr   # Writes _KPRCB pointer to CurrentPrcb field
+        kpcr_obj.Self = kpcr_addr           # Writes _KPCR pointer to Self field
+        kpcr_obj.Prcb = kprcb_addr          # Writes _KPRCB pointer to Prcb field
+
+        # Writes KPCR pointer into GS:[0x18]
+        self.ql.mem.write_ptr(0x6000018, kpcr_addr)
+        
+        # @NOTE: Tests for writing structure pointers.
+        # Please don't remove.
+        self.ql.log.warn(f"KPCR CurrentPrcb: {kpcr_obj.CurrentPrcb:x}")
+        self.ql.log.warn(f"KPCR Self: {kpcr_obj.Self:x}")
+        self.ql.log.warn(f"KPCR Prcb: {kpcr_obj.Prcb:x}")
+
+        self.ql.os.KPCR = kpcr_obj
+
+    def init_kthread(self):
+        '''
+        Initialisation function for KTHREAD structure. 
+        
+        This structures pointer should always be at gs:[0x188]
+        '''
+
+        sysconf = self.ql.os.profile['SYSTEM']
+        osconf  = self.ql.os.profile[f'OS{self.ql.arch.bits}']
+
+        kthread_addr = osconf.getint('KTHREAD')
+        kthread_struct = KTHREAD
+        self.ql.mem.map(kthread_addr, self.ql.mem.align_up(kthread_struct.sizeof()), info='[kthread]')
+
+        # Initialize an instance with key fields
+        kthread_obj = kthread_struct.volatile_ref(self.ql.mem, kthread_addr)
```

Comment:
This structure is initialized with magic numbers whose origin is unclear.
Is there a place [a Qilign object] we can take the values from? Consistency is key.

Reply:
I actually referred to the `kthread_obj` fields values: 11, 0x1000, 0x1500, 0x2000?